task/Update.js

```diff
@@ -54,97 +68,265 @@ async function update() {
+    global.hasFailures = true;
+    postSlackMessage('OSM data update failed, using previous version :boom:');
+  }
 
-    if (global.hasFailures) {
+  await start('gtfs:update');
+
+  process.stdout.write('Build routing graph\n');
+  await start('router:buildGraph');
+
+  if (process.env.SKIPPED_SITES === 'all' || process.env.SKIP_OTP_TESTS) {
+    process.stdout.write('Skipping all tests\n');
+  } else {
+    process.stdout.write('Test the newly built graph with OTPQA\n');
+    execFileSync('./test.sh', [], { stdio: [0, 1, 2] });
+  }
+
+  const logFile = 'failed_feeds.txt';
+  if (fs.existsSync(logFile)) {
+    // testing detected routing problems
+    global.hasFailures = true;
+
+    global.failedFeeds = fs.readFileSync(logFile, 'utf8'); // comma separated list of feed ids. No newline at end!
+    fs.unlinkSync(logFile); // cleanup for local use
+
+    if (global.failedFeeds.split(',').length > MAX_GTFS_FALLBACK) {
+      updateSlackMessage(
+        'Aborting the data update because too many quality tests failed :boom:',
+      );
+      process.exit(1);
+    }
+
+    postSlackMessage(
+      `GTFS packages ${global.failedFeeds} rejected, using fallback to current data`,
+    );
+    // use seed packages for failed feeds
+    await start('gtfs:fallback');
+
+    // rebuild the graph
+    process.stdout.write('Rebuild graph using fallback data\n');
+    await start('router:buildGraph');
+  }
+
+  const date = getDateString();
+
+  global.storageDirName = `${process.env.DOCKER_TAG}/${date}/${name}`;
+
+  process.stdout.write('Uploading data to storage\n');
+  await start('router:store');
+
+  process.stdout.write('Build and deploy Docker images\n');
+  execFileSync('./otp-data-server/deploy.sh', [date], {
+    stdio: [0, 1, 2],
+    env: {
+      OTP_TAG: process.env.OTP_TAG,
+      OTP_GRAPH_DIR: global.storageDirName,
+      ROUTER_NAME: process.env.ROUTER_NAME,
+      ORG: process.env.ORG,
+      DOCKER_TAG: process.env.DOCKER_TAG,
+      DOCKER_USER: process.env.DOCKER_USER,
+      DOCKER_AUTH: process.env.DOCKER_AUTH,
+    },
+  });
+  execFileSync('./opentripplanner/deploy-otp.sh', [date], {
+    stdio: [0, 1, 2],
+    env: {
+      OTP_TAG: process.env.OTP_TAG,
+      OTP_GRAPH_DIR: global.storageDirName,
+      ROUTER_NAME: process.env.ROUTER_NAME,
+      ORG: process.env.ORG,
+      DOCKER_TAG: process.env.DOCKER_TAG,
+      DOCKER_USER: process.env.DOCKER_USER,
+      DOCKER_AUTH: process.env.DOCKER_AUTH,
+    },
+  });
+
+  if (!process.env.NOCLEANUP) {
+    process.stdout.write('Remove oldest data versions from storage\n');
+    await start('storage:cleanup');
+  }
+
+  if (global.hasFailures) {
+    updateSlackMessage(
+      `${name} data updated, but partially falling back to older data :boom:`,
+    );
+  } else {
+    updateSlackMessage(`${name} data updated :white_check_mark:`);
+  }
+}
+
+/**
+ * This function builds the graph from prebuilt street graph data.
+ */
+async function buildWithPrebuiltStreetGraph(name) {
```

Comment:
There is probably quite a bit of same code here as in the regular build. That could be extracted into a method/methods.

Reply:
I extracted a lot of methods